wgengine/magicsock/magicsock.go

```diff
@@ -2233,11 +2227,17 @@ func (c *Conn) handlePingLocked(dm *disco.Ping, src netip.AddrPort, di *discoInf
 	var numNodes int
 	if isDerp {
 		if ep, ok := c.peerMap.endpointForNodeKey(derpNodeSrc); ok {
+			if ep.isDuplicatePing(dm.TxID) {
+				return
+			}
 			ep.addCandidateEndpoint(src)
```

Comment:
this added an extra Lock/Unlock. maybe that's a drop in the bucket, but you could pass dm.TxID to `addCandidateEndpoint` (or some new name or new variant) that does the same work as `addCandidateEndpoint` but returns some value (new bool or sentinel error) that says it was a dup and not change the number of lock operations.

Reply:
I ended up plumbing the TxID down into `addCandidateEndpoint()` and tracking it on `endpointState` where there are other ping-related states.